components/places/src/bookmark_sync/store.rs

```diff
@@ -129,87 +136,398 @@ impl<'a> BookmarksStore<'a> {
+            |chunk, _| -> Result<()> {
+                self.interruptee.err_if_interrupted()?;
+                self.db.execute(
+                    &format!(
+                        "UPDATE moz_bookmarks SET
+                             syncChangeCounter = 1
+                         WHERE guid IN ({})",
+                        sql_support::repeat_sql_vars(chunk.len()),
+                    ),
+                    chunk,
+                )?;
+                Ok(())
+            },
+        )?;
+
+        log::debug!("Flagging applied remote items as merged");
+        sql_support::each_chunk_mapped(
+            &ops.set_remote_merged,
+            |op| op.guid().as_str(),
+            |chunk, _| -> Result<()> {
+                self.interruptee.err_if_interrupted()?;
+                self.db.execute(
+                    &format!(
+                        "UPDATE moz_bookmarks_synced SET
+                             needsMerge = 0
+                         WHERE guid IN ({})",
+                        sql_support::repeat_sql_vars(chunk.len()),
+                    ),
+                    chunk,
+                )?;
+                Ok(())
+            },
+        )?;
+
+        Ok(())
+    }
 
-        // `itemsToMerge` is a view, so "deleting" from it fires the
-        // `insertNewLocalItems` and `updateExistingLocalItems`
-        // triggers instead.
-        self.db.execute_batch("DELETE FROM itemsToMerge")?;
+    fn apply_remote_items(&self, now: Timestamp) -> Result<()> {
+        log::debug!("Removing old tags");
+        self.interruptee.err_if_interrupted()?;
+        self.db.execute_batch(
+            "DELETE FROM moz_tags_relation
+             WHERE place_id IN (SELECT oldPlaceId FROM itemsToApply
+                                WHERE oldPlaceId NOT NULL) OR
+                   place_id IN (SELECT newPlaceId FROM itemsToApply
+                                WHERE newPlaceId NOT NULL)",
+        )?;
 
-        // `structureToMerge` is also a view, so "deleting" from it fires the
-        // `updateLocalStructure` trigger.
-        self.db.execute_batch("DELETE FROM structureToMerge")?;
+        // Insert and update items, using -1 for new items' parent IDs and
+        // positions. We'll update these later, when we apply the new local
+        // structure. This is a full table scan on `itemsToApply`. The no-op
+        // `WHERE` clause is necessary to avoid a parsing ambiguity.
+        log::debug!("Upserting new items");
+        self.interruptee.err_if_interrupted()?;
+        self.db.execute_batch(&format!(
+            "INSERT INTO moz_bookmarks(id, guid, parent, position, type, fk,
+                                       title,
+                                       dateAdded,
+                                       lastModified, syncStatus,
+                                       syncChangeCounter)
+             SELECT localId, mergedGuid, 1, -1,
+                    (CASE
```

Comment:
This big ol' `CASE` could probably be factored out into a `BookmarkTypeFragment` struct...

Along with a comment that unknown kinds in `itemsToApply`, if they somehow sneak through, will cause a `NOT NULL` constraint violation on `moz_bookmarks.type`.